File: java/src/museo/Restauracion.java
package museo;

import java.time.LocalDate;
import java.util.Optional;
import java.util.concurrent.atomic.AtomicInteger;

public class Restauracion {

    private static final AtomicInteger ID_SEQ = new AtomicInteger(0);

    private final int idRestauracion;
    private final ObraArte obra;
    private final String tipoRestauracion;
    private final LocalDate fechaInicio;
    private LocalDate fechaFin;
    private final String motivo;

    public Restauracion(
            ObraArte obra,
            String tipoRestauracion,
            LocalDate fechaInicio,
            LocalDate fechaFin,
            String motivo
    ) {
        this.idRestauracion = ID_SEQ.incrementAndGet();
        this.obra = obra;
        this.tipoRestauracion = tipoRestauracion;
        this.fechaInicio = fechaInicio;
        this.fechaFin = fechaFin;
        this.motivo = motivo;
    }

    public int getIdRestauracion() {
        return idRestauracion;
    }

    public ObraArte getObra() {
        return obra;
    }

    public String getTipoRestauracion() {
        return tipoRestauracion;
    }

    public LocalDate getFechaInicio() {
        return fechaInicio;
    }

    public Optional<LocalDate> getFechaFin() {
        return Optional.ofNullable(fechaFin);
    }

    public String getMotivo() {
        return motivo;
    }

    public void iniciar() {
        obra.enviarARestauracion();
    }

    public void finalizar() {
        this.fechaFin = LocalDate.now();
        obra.marcarExpuesta();
    }

    public boolean estaActiva() {
        return fechaFin == null;
    }
}

File: javascript/src/museo/restauracion.js
import { hoyUTC } from "./fechasUtil.js";

let _idSeq = 0;

export class Restauracion {
  /**
   * @param {import('./obraArte.js').ObraArte} obra
   * @param {string} tipoRestauracion
   * @param {Date} fechaInicio
   * @param {Date | null} fechaFin
   * @param {string} motivo
   */
  constructor(obra, tipoRestauracion, fechaInicio, fechaFin, motivo) {
    this.idRestauracion = ++_idSeq;
    this.obra = obra;
    this.tipoRestauracion = tipoRestauracion;
    this.fechaInicio = fechaInicio;
    this.fechaFin = fechaFin;
    this.motivo = motivo;
  }

  getIdRestauracion() {
    return this.idRestauracion;
  }

  getObra() {
    return this.obra;
  }

  getTipoRestauracion() {
    return this.tipoRestauracion;
  }

  getFechaInicio() {
    return this.fechaInicio;
  }

  /** @returns {Date | null} */
  getFechaFin() {
    return this.fechaFin;
  }

  getMotivo() {
    return this.motivo;
  }

  iniciar() {
    this.obra.enviarARestauracion();
  }

  finalizar(fechaFin = hoyUTC()) {
    this.fechaFin = fechaFin;
    this.obra.marcarExpuesta();
  }

  estaActiva() {
    return this.fechaFin == null;
  }
}
